File: dart/test/root_rule_selection_admission_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:linkedspec_dart/linkedspec_dart.dart';
import 'package:linkedspec_dart/src/cli/primary_cli.dart';
import 'package:test/test.dart';

final Map<String, Object?> _contract = _object(
  jsonDecode(
    File(
      '../capability_conformance/root_rule_selection_contract.json',
    ).readAsStringSync(),
  ),
);

const _generatedIdentity = 'root-selection/dart-admission.spec';

const _markedSource = '''Earlier:
 /x/
 E { return("earlier") }

Marked::
 /x/
 E { return("marked") }

Later::
 /x/
 E { return("later") }
''';

const _markerlessSource = '''First:
 /x/
 E { return("first") }

Second:
 /x/
 E { return("second") }
''';

const _traceSource = '''Top::
 /x/ -> Done { return("trace") }

Done::
 /x/
''';

typedef _AdmissionRole = void Function(Map<String, Object?> contract);

void main() {
  test('contract-declared Dart roles execute once and only once', () {
    final roleMap = <String, _AdmissionRole>{
      'neutral_selection': role_neutral_selection,
      'neutral_failures': role_neutral_failures,
      'neutral_strict': role_neutral_strict,
      'native': role_native,
      'loaded': role_loaded,
      'reconstructed': role_reconstructed,
      'generated_direct': role_generated_direct,
      'generated_traced': role_generated_traced,
      'emitted_source_direct': role_emitted_source_direct,
      'emitted_source_traced': role_emitted_source_traced,
      'descriptor': role_descriptor,
      'diagnostic': role_diagnostic,
      'runtime_trace': role_runtime_trace,
      'primary_cli': role_primary_cli,
      'primary_request_trace': role_primary_request_trace,
    };
    final admission = _object(_contract['dart_admission']);
    final declaredRoles = _strings(admission['roles']);
    expect(
      declaredRoles.toSet().length,
      declaredRoles.length,
      reason: 'Dart admission roles must be unique',
    );
    expect(
      roleMap.keys.toSet(),
      declaredRoles.toSet(),
      reason: 'Dart consumer must implement exactly the declared roles',
    );

    final completed = <String>{};
    for (final role in declaredRoles) {
      expect(completed.add(role), isTrue, reason: 'role $role repeated');
      roleMap[role]!(_contract);
    }
    expect(
      completed,
      roleMap.keys.toSet(),
      reason: 'every declared Dart admission role must complete once',
    );
  });
}

void role_neutral_selection(Map<String, Object?> contract) {
  expect(contract['contract_id'], linkedSpecRootRuleSelectionContract);
  for (final caseValue in _objects(contract['selection_cases'])) {
    final id = caseValue['id']! as String;
    final compiled = _compiledForRows(_objects(caseValue['rules']));
    final authoredBefore = [
      for (final label in compiled.compiledRuleOrder)
        (label, compiled.rule(label)!.header.isTop),
    ];
    final selection = compiled.resolveEntryRule(
      caseValue['explicit_selector'] as String?,
    );
    expect(selection.rule.label, caseValue['expected_label'], reason: id);
    expect(
      selection.basis.contractName,
      caseValue['expected_basis'],
      reason: id,
    );
    expect(
      [
        for (final label in compiled.compiledRuleOrder)
          (label, compiled.rule(label)!.header.isTop),
      ],
      authoredBefore,
      reason: '$id rewrote authored identity',
    );
  }
}

void role_neutral_failures(Map<String, Object?> contract) {
  for (final caseValue in _objects(contract['failure_cases'])) {
    final id = caseValue['id']! as String;
    try {
      _compiledForRows(
        _objects(caseValue['rules']),
      ).resolveEntryRule(caseValue['explicit_selector'] as String?);
      fail('$id must reject selection');
    } on EntryRuleSelectionException catch (error) {
      expect(error.code, caseValue['expected_code'], reason: id);
      expect(error.stage, caseValue['expected_stage'], reason: id);
    }
  }
}

void role_neutral_strict(Map<String, Object?> contract) {
  for (final caseValue in _objects(contract['strict_cases'])) {
    final id = caseValue['id']! as String;
    final source = switch (id) {
      'explicit_selection_is_not_reference' => 'A:\n /a/\n\nB:\n /b/\n',
      'marker_selection_is_not_reference' =>
        'Top::\n /x/ -> Child\n\nChild:\n /x/ -> Child\n',
      'closed_reference_cycle_has_no_unused_rules' =>
        'A:\n /a/ -> B\n\nB:\n /b/ -> A\n',
      _ => throw StateError('unhandled strict case $id'),
    };
    final expectedUnused = _strings(caseValue['expected_unused']);
    if (expectedUnused.isEmpty) {
      expect(
        () => validateSpec(parseSpec(source), strictSyntax: true),
        returnsNormally,
        reason: id,
      );
    } else {
      expect(
        () => validateSpec(parseSpec(source), strictSyntax: true),
        throwsA(
          isA<SpecValidationException>().having(
            (error) => error.message,
            'unused labels',
            contains(expectedUnused.join(', ')),
          ),
        ),
        reason: id,
      );
    }
  }
}

void role_native(Map<String, Object?> _) {
  final marked = _engine(_markedSource);
  expect(marked.parse('x').value, 'marked');
  expect(marked.parse('x', topRule: 'Earlier').value, 'earlier');
  expect(marked.parse('x', topRule: 'Later').value, 'later');

  final markerless = _engine(_markerlessSource);
  expect(markerless.parse('x').value, 'first');
  expect(markerless.parse('x', topRule: 'Second').value, 'second');
}

void role_loaded(Map<String, Object?> _) {
  _withScratch('loaded', (scratch) {
    final file = File('${scratch.path}${Platform.pathSeparator}markerless.spec')
      ..writeAsStringSync(_markerlessSource);
    final loaded = loadAndCompileSpec(
      SpecRequest.path(file.path),
      SpecLoadOptions(cwd: scratch),
    );
    expect(loaded.createEngine().parse('x').value, 'first');
    expect(loaded.createEngine().parse('x', topRule: 'Second').value, 'second');
  });
}

void role_reconstructed(Map<String, Object?> _) {
  final normalized = SpecFile.fromJson(
    _object(jsonDecode(jsonEncode(parseSpec(_markedSource).toJson()))),
  );
  final reconstructed = compileSpec(normalized);
  expect(
    [
      for (final label in reconstructed.compiledRuleOrder)
        (label, reconstructed.rule(label)!.header.isTop),
    ],
    [('Earlier', false), ('Marked', true), ('Later', true)],
  );
  final engine = LinkedSpecRuntimeEngine(reconstructed);
  expect(engine.parse('x').value, 'marked');
  expect(engine.parse('x', topRule: 'Earlier').value, 'earlier');
}

void role_generated_direct(Map<String, Object?> _) {
  final marked = _compileSource(_markedSource);
  expect(
    executeGeneratedParserV2(
      marked,
      buildGeneratedRulePlan(marked),
      'x',
      _generatedIdentity,
      topRule: 'Earlier',
    ),
    'earlier',
  );

  final markerless = _compileSource(_markerlessSource);
  expect(
    executeGeneratedParserV2(
      markerless,
      buildGeneratedRulePlan(markerless),
      'x',
      _generatedIdentity,
    ),
    'first',
  );
}

void role_generated_traced(Map<String, Object?> _) {
  _withScratch('generated-traced', (scratch) {
    final compiled = _compileSource(_markedSource);
    final traceFile = File(
      '${scratch.path}${Platform.pathSeparator}generated.trace',
    );
    expect(
      executeGeneratedParserWithTraceV2(
        compiled,
        buildGeneratedRulePlan(compiled),
        'x',
        LinkedSpecTraceConfig(
          level: LinkedSpecTraceLevel.debug,
          traceFile: traceFile.path,
          sinkMode: LinkedSpecTraceSinkMode.route,
          resetFile: true,
        ),
        _generatedIdentity,
        topRule: 'Later',
      ),
      'later',
    );
    final trace = traceFile.readAsStringSync();
    expect(trace, contains('dart_runtime:entry_rule_selection'));
    expect(trace, contains('requested=Later'));
    expect(trace, contains('effective=Later'));
    expect(trace, contains('basis=explicit_selector'));
    expect(trace, contains('rule=Later'));
  });
}

void role_emitted_source_direct(Map<String, Object?> _) {
  final source = emitDartSourceV2(
    _compileSource(_markedSource),
    _generatedIdentity,
  );
  for (final marker in [
    'const linkedspecGeneratedSourceContract =',
    "'linkedspec-generated-source-v2'",
    'const linkedspecGeneratedSourceFormat = 2;',
    'Object? execute(',
    'String? topRule,',
    _generatedIdentity,
  ]) {
    expect(source, contains(marker), reason: 'emitted source omitted $marker');
  }
}

void role_emitted_source_traced(Map<String, Object?> _) {
  final source = emitDartSourceV2(
    _compileSource(_markedSource),
    _generatedIdentity,
  );
  for (final marker in [
    'Object? executeWithTrace(',
    'LinkedSpecTraceConfig traceConfig',
    'executeGeneratedParserWithTraceV2(',
    'topRule: topRule',
  ]) {
    expect(source, contains(marker), reason: 'emitted source omitted $marker');
  }
}

void role_descriptor(Map<String, Object?> _) {
  final compiled = _compileSource(_markedSource);
  final before = compiled.toDescriptorJson();
  expect(
    LinkedSpecRuntimeEngine(compiled).parse('x', topRule: 'Earlier').value,
    'earlier',
  );
  final after = compiled.toDescriptorJson();
  expect(after, before);
  final meta = _object(after['meta']);
  expect(meta['entry_rule_contract'], linkedSpecRootRuleSelectionContract);
  expect(meta['definition_order'], ['Earlier', 'Marked', 'Later']);
  expect(meta, isNot(contains('entry_rule')));
  expect(meta, isNot(contains('selected_entry_rule')));
  final spec = _object(after['spec']);
  expect(_object(_object(spec['Earlier'])['meta'])['is_top'], isFalse);
  expect(_object(_object(spec['Marked'])['meta'])['is_top'], isTrue);
}

void role_diagnostic(Map<String, Object?> _) {
  try {
    _engine(_markedSource).parse('x', topRule: 'Missing');
    fail('unknown selector must reject before execution');
  } on RuntimeInterpreterException catch (error) {
    expect(error.diagnostic?.code, 'entry_rule_not_found');
    expect(error.diagnostic?.stage, 'select_entry_rule');
    expect(error.diagnostic?.entryRule, 'Missing');
  }

  final empty = LinkedSpecRuntimeEngine(
    compileSpec(const SpecFile(rules: []), validateSource: false),
  );
  try {
    empty.parse('', topRule: 'Missing');
    fail('zero rules must reject before explicit selection');
  } on RuntimeInterpreterException catch (error) {
    expect(error.diagnostic?.code, 'no_rules_defined');
    expect(error.diagnostic?.stage, 'validate_spec');
  }

  final compiled = _compileSource(_markedSource);
  try {
    executeGeneratedParserV2(
      compiled,
      buildGeneratedRulePlan(compiled),
      'x',
      _generatedIdentity,
      topRule: 'Missing',
      actualContract: 'linkedspec-generated-source-v1',
    );
    fail('stale generated contract must reject before selection');
  } on GeneratedSourceException catch (error) {
    expect(error.stage, GeneratedSourceStage.validateGeneratedPlan);
    expect(
      error.code,
      GeneratedSourceCode.generatedSourceContractVersionMismatch,
    );
  }
}

void role_runtime_trace(Map<String, Object?> _) {
  const failureSource = '''Earlier:
 /x/
 E { return(not(true, false)) }

Marked::
 /x/
 E { return("marked") }
''';
  _withScratch('runtime-trace', (scratch) {
    final compiled = _compileSource(failureSource);
    final traceFile = File(
      '${scratch.path}${Platform.pathSeparator}failure.trace',
    );
    try {
      executeGeneratedParserWithTraceV2(
        compiled,
        buildGeneratedRulePlan(compiled),
        'x',
        LinkedSpecTraceConfig(
          level: LinkedSpecTraceLevel.debug,
          traceFile: traceFile.path,
          sinkMode: LinkedSpecTraceSinkMode.route,
          resetFile: true,
        ),
        _generatedIdentity,
        topRule: 'Earlier',
      );
      fail('selected rule runtime failure must retain effective identity');
    } on GeneratedSourceException catch (error) {
      expect(error.stage, GeneratedSourceStage.executeGenerated);
      expect(error.code, GeneratedSourceCode.generatedExecutionFailed);
      expect(error.ruleLabel, 'Earlier');
      expect(error.handlerFamily, 'default');
    }
    final trace = traceFile.readAsStringSync();
    expect(trace, contains('requested=Earlier'));
    expect(trace, contains('effective=Earlier'));
    expect(trace, contains('basis=explicit_selector'));
    expect(trace, contains('rule=Earlier'));
  });
}

void role_primary_cli(Map<String, Object?> _) {
  for (final route in [
    (_markedSource, <String>[], '"marked"\n'),
    (_markerlessSource, <String>[], '"first"\n'),
    (_markedSource, ['--top-rule', 'Earlier'], '"earlier"\n'),
  ]) {
    final output = runLinkedSpecDartPrimaryCli([
      '--inline-spec',
      route.$1,
      '--input',
      'x',
      ...route.$2,
    ]);
    expect(output.exitCode, 0);
    expect(output.stdoutBytes, utf8.encode(route.$3));
    expect(output.stderrBytes, isEmpty);
  }

  final unknown = runLinkedSpecDartPrimaryCli(const [
    '--inline-spec',
    _markedSource,
    '--input',
    'x',
    '--top-rule',
    'Missing',
  ]);
  expect(unknown.exitCode, 1);
  expect(unknown.stdoutBytes, isEmpty);
  expect(
    unknown.stderrBytes,
    utf8.encode('linkedspec: parser invocation failed\n'),
  );
}

void role_primary_request_trace(Map<String, Object?> _) {
  final expectedDefault = File(
    '../cli_conformance/cases/trace/medium_stdout.txt',
  ).readAsBytesSync();
  final defaultOutput = runLinkedSpecDartPrimaryCli(const [
    '--inline-spec',
    _traceSource,
    '--input',
    'x',
    '--trace',
    'medium',
  ]);
  expect(defaultOutput.exitCode, 0);
  expect(defaultOutput.stdoutBytes, expectedDefault);
  expect(defaultOutput.stderrBytes, isEmpty);

  _withScratch('primary-request-trace', (scratch) {
    final traceFile = File('${scratch.path}${Platform.pathSeparator}trace.log')
      ..writeAsStringSync('stale trace\n');
    final explicitOutput = runLinkedSpecDartPrimaryCli(
      const [
        '--inline-spec',
        _traceSource,
        '--input',
        'x',
        '--top-rule',
        'Top\nInjected',
        '--trace',
        'medium',
        '--trace-file',
        'trace.log',
        '--trace-mode',
        'route',
        '--trace-reset',
      ],
      workingDirectory: scratch,
      repositoryRoot: Directory('..'),
    );
    expect(explicitOutput.exitCode, 1);
    expect(explicitOutput.stdoutBytes, isEmpty);
    expect(
      explicitOutput.stderrBytes,
      utf8.encode('linkedspec: parser invocation failed\n'),
    );
    expect(
      traceFile.readAsBytesSync(),
      File(
        '../cli_conformance/cases/trace/failure_invoke_escaped_medium.txt',
      ).readAsBytesSync(),
    );
  });
}

CompiledSpec _compileSource(String source) {
  final parsed = parseSpec(source);
  validateSpec(parsed);
  return compileSpec(parsed, validateSource: false);
}

CompiledSpec _compiledForRows(List<Map<String, Object?>> rows) {
  if (rows.isEmpty) {
    return compileSpec(const SpecFile(rules: []), validateSource: false);
  }
  final source = rows
      .map((row) {
        final label = row['label']! as String;
        final separator = row['authored_is_top'] == true ? '::' : ':';
        return '$label$separator\n /x/\n E { return("$label") }\n';
      })
      .join('\n');
  return _compileSource(source);
}

LinkedSpecRuntimeEngine _engine(String source) =>
    LinkedSpecRuntimeEngine(_compileSource(source));

T _withScratch<T>(String label, T Function(Directory scratch) body) {
  final scratch = Directory.systemTemp.createTempSync(
    'linkedspec-dart-root-admission-$label-',
  );
  try {
    return body(scratch);
  } finally {
    if (scratch.existsSync()) {
      scratch.deleteSync(recursive: true);
    }
  }
}

Map<String, Object?> _object(Object? value) =>
    Map<String, Object?>.from(value! as Map);

List<Map<String, Object?>> _objects(Object? value) => [
  for (final row in value! as List) _object(row),
];

List<String> _strings(Object? value) => [
  for (final item in value! as List) item! as String,
];

File: julia/test/root_rule_selection_admission_test.jl
const JULIA_ROOT_ADMISSION_CONTRACT = JSON3.read(
    read(
        joinpath(REPO_ROOT, "capability_conformance", "root_rule_selection_contract.json"),
        String,
    ),
    Dict{String,Any},
)

const JULIA_ROOT_ADMISSION_IDENTITY = "root-selection/julia-admission.spec"

# Entry lifecycle returns make the selected rule observable before matching can
# make exit lifecycle appear equivalent.
const JULIA_ROOT_ADMISSION_MARKED_SOURCE = raw"""
Earlier:
 /x/
 I { return("earlier") }

Marked::
 /x/
 I { return("marked") }

Later::
 /x/
 I { return("later") }
"""

const JULIA_ROOT_ADMISSION_MARKERLESS_SOURCE = raw"""
First:
 /x/
 I { return("first") }

Second:
 /x/
 I { return("second") }
"""

# This source is shared byte-for-byte with the primary CLI trace manifest.
const JULIA_ROOT_ADMISSION_TRACE_SOURCE = raw"""Top::
 /x/ -> Done { return("trace") }

Done::
 /x/
"""

function _julia_root_admission_compile(source)
    spec = parse_spec(source)
    validate_spec(spec)
    return compile_spec(spec; validate_source = false)
end

function _julia_root_admission_compiled_for_rows(rows)
    if isempty(rows)
        return compile_spec(SpecFile(rules = Rule[]); validate_source = false)
    end
    source = join(
        [
            begin
                label = String(row["label"])
                separator = row["authored_is_top"] ? "::" : ":"
                "$label$separator\n /x/\n I { return(\"$label\") }\n"
            end for row in rows
        ],
        "\n",
    )
    return _julia_root_admission_compile(source)
end

function _julia_root_admission_error(call)
    try
        call()
    catch error
        return error
    end
    return nothing
end

_julia_root_admission_latest_binding(owner::Module, name::Symbol) =
    Base.invokelatest(() -> Core.getglobal(owner, name))

function _julia_root_admission_with_emitted_parser(body, source, label)
    compiled = _julia_root_admission_compile(source)
    mktempdir() do scratch
        generated_path = joinpath(scratch, "generated_parser.jl")
        write(generated_path, emit_julia_source_v2(compiled, JULIA_ROOT_ADMISSION_IDENTITY))
        host = Module(gensym(:JuliaRootAdmissionHost))
        Base.include(host, generated_path)
        parser = Base.invokelatest(() -> getfield(host, :LinkedSpecGeneratedParser))
        return body(parser, label)
    end
end

function role_neutral_selection(contract)
    @test contract["contract_id"] == ENTRY_RULE_CONTRACT_ID
    for case_value in contract["selection_cases"]
        case_id = String(case_value["id"])
        compiled = _julia_root_admission_compiled_for_rows(case_value["rules"])
        authored_before = [
            (label, compiled_rule(compiled, label).header.is_top)
            for label in compiled.compiled_rule_order
        ]
        selection = resolve_entry_rule(compiled, case_value["explicit_selector"])
        @test selection.rule.label == case_value["expected_label"]
        @test entry_rule_selection_basis_name(selection.basis) ==
            case_value["expected_basis"]
        @test [
            (label, compiled_rule(compiled, label).header.is_top)
            for label in compiled.compiled_rule_order
        ] == authored_before
        @test !isempty(case_id)
    end
end

function role_neutral_failures(contract)
    for case_value in contract["failure_cases"]
        compiled = _julia_root_admission_compiled_for_rows(case_value["rules"])
        error = _julia_root_admission_error() do
            resolve_entry_rule(compiled, case_value["explicit_selector"])
        end
        @test error isa EntryRuleSelectionException
        if error isa EntryRuleSelectionException
            @test error.code == case_value["expected_code"]
            @test error.stage == case_value["expected_stage"]
        end
    end
end

function role_neutral_strict(contract)
    sources = Dict(
        "explicit_selection_is_not_reference" => "A:\n /a/\n\nB:\n /b/\n",
        "marker_selection_is_not_reference" =>
            "Top::\n /x/ -> Child\n\nChild:\n /x/ -> Child\n",
        "closed_reference_cycle_has_no_unused_rules" =>
            "A:\n /a/ -> B\n\nB:\n /b/ -> A\n",
    )
    for case_value in contract["strict_cases"]
        case_id = String(case_value["id"])
        spec = parse_spec(sources[case_id])
        expected_unused = String[String(label) for label in case_value["expected_unused"]]
        if case_value["explicit_selector"] !== nothing
            resolve_entry_rule(compile_spec(spec), case_value["explicit_selector"])
        end
        if isempty(expected_unused)
            @test validate_spec(spec; strict_syntax = true) === nothing
        else
            error = _julia_root_admission_error() do
                validate_spec(spec; strict_syntax = true)
            end
            @test error isa SpecValidationException
            if error isa SpecValidationException
                @test error.message ==
                    "unused rule(s) in strict mode: $(join(expected_unused, ", "))"
            end
        end
    end
end

function role_native(_contract)
    marked = LinkedSpecRuntimeEngine(
        _julia_root_admission_compile(JULIA_ROOT_ADMISSION_MARKED_SOURCE),
    )
    @test runtime_parse(marked, "x").value == "marked"
    @test runtime_parse(marked, "x"; top_rule = "Earlier").value == "earlier"
    @test runtime_parse(marked, "x"; top_rule = "Later").value == "later"

    markerless = LinkedSpecRuntimeEngine(
        _julia_root_admission_compile(JULIA_ROOT_ADMISSION_MARKERLESS_SOURCE),
    )
    @test runtime_parse(markerless, "x").value == "first"
    @test runtime_parse(markerless, "x"; top_rule = "Second").value == "second"
end

function role_loaded(_contract)
    mktempdir() do scratch
        path = joinpath(scratch, "markerless.spec")
        write(path, JULIA_ROOT_ADMISSION_MARKERLESS_SOURCE)
        loaded = load_and_compile_spec(path_spec_request(path), SpecLoadOptions(scratch))
        engine = create_engine(loaded)
        @test runtime_parse(engine, "x").value == "first"
        @test runtime_parse(engine, "x"; top_rule = "Second").value == "second"
    end
end

function role_reconstructed(_contract)
    normalized_json = JSON3.read(
        JSON3.write(to_json(parse_spec(JULIA_ROOT_ADMISSION_MARKED_SOURCE))),
    )
    compiled = compile_spec(from_json(SpecFile, normalized_json))
    @test [
        (label, compiled_rule(compiled, label).header.is_top)
        for label in compiled.compiled_rule_order
    ] == [("Earlier", false), ("Marked", true), ("Later", true)]
    engine = LinkedSpecRuntimeEngine(compiled)
    @test runtime_parse(engine, "x").value == "marked"
    @test runtime_parse(engine, "x"; top_rule = "Earlier").value == "earlier"
end

function role_generated_direct(_contract)
    marked = _julia_root_admission_compile(JULIA_ROOT_ADMISSION_MARKED_SOURCE)
    @test execute_generated_parser_v2(
        marked,
        build_generated_rule_plan(marked),
        "x",
        JULIA_ROOT_ADMISSION_IDENTITY;
        top_rule = "Earlier",
    ) == "earlier"

    markerless = _julia_root_admission_compile(JULIA_ROOT_ADMISSION_MARKERLESS_SOURCE)
    @test execute_generated_parser_v2(
        markerless,
        build_generated_rule_plan(markerless),
        "x",
        JULIA_ROOT_ADMISSION_IDENTITY,
    ) == "first"
end

function role_generated_traced(_contract)
    mktempdir() do scratch
        compiled = _julia_root_admission_compile(JULIA_ROOT_ADMISSION_MARKED_SOURCE)
        trace_path = joinpath(scratch, "generated.trace")
        @test execute_generated_parser_with_trace_v2(
            compiled,
            build_generated_rule_plan(compiled),
            "x",
            LinkedSpecTraceConfig(
                level = LinkedSpecTraceDebug,
                trace_file = trace_path,
                sink_mode = LinkedSpecTraceRoute,
                reset_file = true,
            ),
            JULIA_ROOT_ADMISSION_IDENTITY;
            top_rule = "Later",
        ) == "later"
        trace = read(trace_path, String)
        @test occursin("julia_runtime:entry_rule_selection", trace)
        @test occursin("requested=Later", trace)
        @test occursin("effective=Later", trace)
        @test occursin("basis=explicit_selector", trace)
        @test occursin("rule=Later", trace)
    end
end

function role_emitted_source_direct(_contract)
    _julia_root_admission_with_emitted_parser(
        JULIA_ROOT_ADMISSION_MARKED_SOURCE,
        "Earlier",
    ) do parser, label
        execute = _julia_root_admission_latest_binding(parser, :execute)
        @test Base.invokelatest(execute, "x") == "marked"
        @test Base.invokelatest(execute, "x"; top_rule = label) == "earlier"
    end
end

function role_emitted_source_traced(_contract)
    _julia_root_admission_with_emitted_parser(
        JULIA_ROOT_ADMISSION_MARKERLESS_SOURCE,
        "Second",
    ) do parser, label
        output = IOBuffer()
        result = Base.invokelatest(
            _julia_root_admission_latest_binding(parser, :execute_with_trace),
            "x",
            trace_config_enabled("low");
            top_rule = label,
            stdout_io = output,
        )
        @test result == "second"
        trace = String(take!(output))
        @test occursin("julia_runtime:entry_rule_selection", trace)
        @test occursin("requested=Second", trace)
        @test occursin("effective=Second", trace)
        @test occursin("basis=explicit_selector", trace)
    end
end

function role_descriptor(_contract)
    compiled = _julia_root_admission_compile(JULIA_ROOT_ADMISSION_MARKED_SOURCE)
    before = to_descriptor_json(compiled)
    @test runtime_parse(
        LinkedSpecRuntimeEngine(compiled),
        "x";
        top_rule = "Earlier",
    ).value == "earlier"
    after = to_descriptor_json(compiled)
    @test after == before
    @test after["meta"]["entry_rule_contract"] == ENTRY_RULE_CONTRACT_ID
    @test after["meta"]["definition_order"] == ["Earlier", "Marked", "Later"]
    @test !haskey(after["meta"], "entry_rule")
    @test !haskey(after["meta"], "selected_entry_rule")
    @test after["spec"]["Earlier"]["meta"]["is_top"] === false
    @test after["spec"]["Marked"]["meta"]["is_top"] === true
    @test after["spec"]["Later"]["meta"]["is_top"] === true
end

function role_diagnostic(_contract)
    engine = LinkedSpecRuntimeEngine(
        _julia_root_admission_compile(JULIA_ROOT_ADMISSION_MARKED_SOURCE),
    )
    unknown = _julia_root_admission_error() do
        runtime_parse(engine, "x"; top_rule = "Missing")
    end
    @test unknown isa RuntimeInterpreterException
    if unknown isa RuntimeInterpreterException
        @test unknown.diagnostic.code == "entry_rule_not_found"
        @test unknown.diagnostic.stage == "select_entry_rule"
        @test unknown.diagnostic.entry_rule == "Missing"
    end

    empty = LinkedSpecRuntimeEngine(
        compile_spec(SpecFile(rules = Rule[]); validate_source = false),
    )
    zero = _julia_root_admission_error() do
        runtime_parse(empty, ""; top_rule = "Missing")
    end
    @test zero isa RuntimeInterpreterException
    if zero isa RuntimeInterpreterException
        @test zero.diagnostic.code == "no_rules_defined"
        @test zero.diagnostic.stage == "validate_spec"
    end

    compiled = _julia_root_admission_compile(JULIA_ROOT_ADMISSION_MARKED_SOURCE)
    stale = _julia_root_admission_error() do
        execute_generated_parser_v2(
            compiled,
            build_generated_rule_plan(compiled),
            "x",
            JULIA_ROOT_ADMISSION_IDENTITY;
            top_rule = "Missing",
            actual_contract = "linkedspec-generated-source-v1",
        )
    end
    @test stale isa GeneratedSourceException
    if stale isa GeneratedSourceException
        @test stale.stage == ValidateGeneratedPlanStage
        @test stale.code == GeneratedSourceContractVersionMismatchCode
    end
end

function role_runtime_trace(_contract)
    failure_source = raw"""
Earlier:
 /x/
 I { return(not(true, false)) }

Marked::
 /x/
 I { return("marked") }
"""
    mktempdir() do scratch
        compiled = _julia_root_admission_compile(failure_source)
        trace_path = joinpath(scratch, "failure.trace")
        failure = _julia_root_admission_error() do
            execute_generated_parser_with_trace_v2(
                compiled,
                build_generated_rule_plan(compiled),
                "x",
                LinkedSpecTraceConfig(
                    level = LinkedSpecTraceDebug,
                    trace_file = trace_path,
                    sink_mode = LinkedSpecTraceRoute,
                    reset_file = true,
                ),
                JULIA_ROOT_ADMISSION_IDENTITY;
                top_rule = "Earlier",
            )
        end
        @test failure isa GeneratedSourceException
        if failure isa GeneratedSourceException
            @test failure.stage == ExecuteGeneratedStage
            @test failure.code == GeneratedExecutionFailedCode
            @test failure.rule_label == "Earlier"
            @test failure.handler_family == "default"
        end
        trace = read(trace_path, String)
        @test occursin("requested=Earlier", trace)
        @test occursin("effective=Earlier", trace)
        @test occursin("basis=explicit_selector", trace)
        @test occursin("rule=Earlier", trace)
    end
end

function role_primary_cli(_contract)
    routes = (
        (JULIA_ROOT_ADMISSION_MARKED_SOURCE, String[], "\"marked\"\n"),
        (JULIA_ROOT_ADMISSION_MARKERLESS_SOURCE, String[], "\"first\"\n"),
        (
            JULIA_ROOT_ADMISSION_MARKED_SOURCE,
            ["--top-rule", "Earlier"],
            "\"earlier\"\n",
        ),
    )
    for (source, extra, expected) in routes
        output = IOBuffer()
        error_output = IOBuffer()
        @test run_cli(
            ["--inline-spec", source, "--input", "x", extra...];
            io = output,
            err = error_output,
        ) == 0
        @test String(take!(output)) == expected
        @test isempty(String(take!(error_output)))
    end

    output = IOBuffer()
    error_output = IOBuffer()
    @test run_cli(
        [
            "--inline-spec",
            JULIA_ROOT_ADMISSION_MARKED_SOURCE,
            "--input",
            "x",
            "--top-rule",
            "Missing",
        ];
        io = output,
        err = error_output,
    ) == 1
    @test isempty(String(take!(output)))
    @test String(take!(error_output)) == "linkedspec: parser invocation failed\n"
end

function role_primary_request_trace(_contract)
    expected_default = read(
        joinpath(REPO_ROOT, "cli_conformance", "cases", "trace", "medium_stdout.txt"),
    )
    output = IOBuffer()
    error_output = IOBuffer()
    @test run_cli(
        [
            "--inline-spec",
            JULIA_ROOT_ADMISSION_TRACE_SOURCE,
            "--input",
            "x",
            "--trace",
            "medium",
        ];
        io = output,
        err = error_output,
    ) == 0
    @test take!(output) == expected_default
    @test isempty(take!(error_output))

    mktempdir() do scratch
        trace_path = joinpath(scratch, "trace.log")
        write(trace_path, "stale trace\n")
        output = IOBuffer()
        error_output = IOBuffer()
        status = cd(scratch) do
            run_cli(
                [
                    "--inline-spec",
                    JULIA_ROOT_ADMISSION_TRACE_SOURCE,
                    "--input",
                    "x",
                    "--top-rule",
                    "Top\nInjected",
                    "--trace",
                    "medium",
                    "--trace-file",
                    "trace.log",
                    "--trace-mode",
                    "route",
                    "--trace-reset",
                ];
                io = output,
                err = error_output,
            )
        end
        @test status == 1
        @test isempty(take!(output))
        @test String(take!(error_output)) == "linkedspec: parser invocation failed\n"
        @test read(trace_path) == read(
            joinpath(
                REPO_ROOT,
                "cli_conformance",
                "cases",
                "trace",
                "failure_invoke_escaped_medium.txt",
            ),
        )
    end
end

@testset "Contract-declared Julia root-selection roles execute once" begin
    role_map = Dict{String,Function}(
        "neutral_selection" => role_neutral_selection,
        "neutral_failures" => role_neutral_failures,
        "neutral_strict" => role_neutral_strict,
        "native" => role_native,
        "loaded" => role_loaded,
        "reconstructed" => role_reconstructed,
        "generated_direct" => role_generated_direct,
        "generated_traced" => role_generated_traced,
        "emitted_source_direct" => role_emitted_source_direct,
        "emitted_source_traced" => role_emitted_source_traced,
        "descriptor" => role_descriptor,
        "diagnostic" => role_diagnostic,
        "runtime_trace" => role_runtime_trace,
        "primary_cli" => role_primary_cli,
        "primary_request_trace" => role_primary_request_trace,
    )
    admission = JULIA_ROOT_ADMISSION_CONTRACT["julia_admission"]
    declared_roles = String[String(role) for role in admission["roles"]]
    @test length(unique(declared_roles)) == length(declared_roles)
    @test Set(keys(role_map)) == Set(declared_roles)

    completed = Set{String}()
    for role in declared_roles
        @test !(role in completed)
        push!(completed, role)
        role_map[role](JULIA_ROOT_ADMISSION_CONTRACT)
    end
    @test completed == Set(keys(role_map))
end
